### PR TITLE
[optim][sgd] default to foreach when CUDA + differentiable=False

### PR DESCRIPTION
**Pull request by janeyx99**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #92730
* #92923

### CODE REVIEW

File: torch/optim/sgd.py [hunk 1]
```diff
@@ -204,8 +204,14 @@ def sgd(params: List[Tensor],
     """
 
     if foreach is None:
-        # Placeholder for more complex foreach logic to be added when value is not set
-        foreach = False
+        # why must we handle momentum_buffer_list separately?
```
Comment (albanD) [inline torch/optim/sgd.py:207]: Is it when tracing or scripting? If the later, than can we just guard this all statement with a is_scripting ?

**Review by albanD — APPROVED** [review]
Sounds good!

File: torch/optim/sgd.py (hunk 1 above)
Comment (janeyx99) [reply torch/optim/sgd.py]: The test passes now (so it is scripting in the test) but should we worry about tracing requirements?

**Pull request closed without merging**